src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/DataSourceUtils.java

```diff
@@ -228,13 +229,84 @@ public static boolean isValidDirectory(final Path p) {
+                    funcotationFactory = DataSourceUtils.createLocatableXsvDataSource(path, properties, annotationOverridesMap, featureInput);
+                    break;
+                case SIMPLE_XSV:
+                    funcotationFactory = DataSourceUtils.createSimpleXsvDataSource(path, properties, annotationOverridesMap);
+                    break;
+                case COSMIC:
+                    funcotationFactory = DataSourceUtils.createCosmicDataSource(path, properties, annotationOverridesMap);
+                    break;
+                case GENCODE:
+                    featureInput = createAndRegisterFeatureInputs(path, properties, funcotatorToolInstance, lookaheadFeatureCachingInBp, GencodeGtfFeature.class);
+                    funcotationFactory = DataSourceUtils.createGencodeDataSource(path, properties, annotationOverridesMap, transcriptSelectionMode, userTranscriptIdSet, featureInput);
+                    break;
+                case VCF:
+                    featureInput = createAndRegisterFeatureInputs(path, properties, funcotatorToolInstance, lookaheadFeatureCachingInBp, VariantContext.class);
+                    funcotationFactory = DataSourceUtils.createVcfDataSource(path, properties, annotationOverridesMap, featureInput);
+                    break;
+                default:
+                    throw new GATKException("Unknown type of DataSourceFuncotationFactory encountered: " + stringType );
+            }
+
+            // Add in our factory:
+            dataSourceFactories.add(funcotationFactory);
+        }
+
+        logger.debug("All Data Sources have been created.");
+        return dataSourceFactories;
+    }
 
+    /**
+     * Create a {@link List} of {@link DataSourceFuncotationFactory} based on meta data on the data sources, overrides, and transcript reporting priority information.
+     * THIS METHOD IS FOR TESTING ONLY!
+     * @param dataSourceMetaData {@link Map} of {@link Path}->{@link Properties} containing metadata about each data source.  Must not be {@code null}.
+     * @param annotationOverridesMap {@link LinkedHashMap} of {@link String}->{@link String} containing any annotation overrides to include in data sources.  Must not be {@code null}.
+     * @param transcriptSelectionMode {@link TranscriptSelectionMode} to use when choosing the transcript for detailed reporting.  Must not be {@code null}.
+     * @param userTranscriptIdSet {@link Set} of {@link String}s containing transcript IDs of interest to be selected for first.  Must not be {@code null}.
+     * @return A {@link List} of {@link DataSourceFuncotationFactory} given the data source metadata, overrides, and transcript reporting priority information.
+     */
+    @VisibleForTesting
+    public static List<DataSourceFuncotationFactory> createDataSourceFuncotationFactoriesForDataSourcesForTesting(
```

Comment:
Is there some way you can avoid duplicating all of the code from `createDataSourceFuncotationFactoriesForDataSources()` in `createDataSourceFuncotationFactoriesForDataSourcesForTesting()` (eg., by factoring out a common helper method)?

I'm also wondering whether you can avoid the need for separate code paths for testing entirely by creating dummy tool instances in your test code. It's not ideal to have test coverage only for test-only codepaths...

Reply:
Fixed.

Reply:
Fixed.